### ISSUE TITLE
Default EditorConfig File

### ISSUE THREAD

**Issue by deinok** [editorconfig/editorconfig#298]
Create a default .editorconfig with the the recomended configurations for every file type.

Ex:

json -> 2 spaces
py -> 4 spaces
etc...

**Comment by jedmao** [comment]
I could imagine a great amount of controversy on this subject. Personally, I only target `package.json` with `indent_size = 2` and let all my other JSON files be tabs.

**Comment by deinok** [comment]
@jedmao Yes, this is exactly the problem, or the benefit

For example Python(PEP8) states 4 spaces of indentation.
For Json there is no convention. But as it is from JavaScript the prefered indentation will be 4.

This Default EditorConfig should only target convention styles. (Probably the json is a bad example)

**Comment by jedmao** [comment]
Can you think of any other language, other than Python, that would have a standard?

**Comment by jedmao** [comment]
So far, it sounds like these would be the only defaults:
```ini
[package.json]
indent_style = space
indent_size = 2

[*.py]
indent_style = space
indent_size = 4
```

**Comment by deinok** [comment]
Java, C#, also i think JavaScript also have some standards

**Comment by deinok** [comment]
What about creating a default file with all the standards?

**Comment by jedmao** [comment]
You would have to track down the standards first. I don't think JavaScript has any set standard.

**Comment by deinok** [comment]
Yep, thats exactly my point. 
A file with the configuration of every extensions that have a standard

**Comment by jedmao** [comment]
I guess I'm asking you to link to the standards for each language that you'd like to see, because I'm feeling like there's not as many standards as you think.

**Comment by deinok** [comment]
Python -> https://www.python.org/dev/peps/pep-0008/

**Comment by deinok** [comment]
JavaScript -> https://www.w3schools.com/js/js_conventions.asp

**Comment by deinok** [comment]
Java -> http://www.oracle.com/technetwork/java/javase/documentation/codeconventions-136091.html#262

**Comment by deinok** [comment]
C# -> https://msdn.microsoft.com/en-us/library/ff926074.aspx#Anchor_1

**Comment by MAPJe71** [comment]
These are links to style guides and conventions, not _standards_ i.e. required by the language IMO.

**Comment by deinok** [comment]
@MAPJe71 Yes, they are conventions, not statements. But still beeing something good have all these conventions in one single file.
The only one will be YAML with its 2 spaces indentation.

**Comment by jedmao** [comment]
@deinok I'm trying to say the same thing as @MAPJe71. These are not standards, but conventions. People get really religious about their choice of conventions, which is why I don't think it has any place in this EditorConfig project, which is very agnostic to user preferene. What you want probably belongs in a separate project or in your own Gist or something.

**Comment by MAPJe71** [comment]
@deinok 
YAML spec 1.2 defines indentation as `zero or more space characters` and for portability tabs should not be used (see [Chapter 6 - Paragraph 1](http://www.yaml.org/spec/1.2/spec.html#Syntax)). YAML files I've seen all use 4 spaces for indentation.

@jedmao totally agree!

**Comment by xuhdev** [comment]
Of course you can always create a project as a collection of popular coding style conventions. Or, even better, we can just do a wiki page, so people can copy and paste easily.

**Comment by treyhunner** [comment]
I like the idea of a project or a wiki page that lists standards as well as conventions.

One more standard I'd add in addition to the above:

```
[Makefile]
indent_style = tab
```

**Comment by ffes** [comment]
Project files for Visual Studio are generated with 2 spaces and Windows line endings (from the top of my head), both the ones in XML and JSON.

YAML needs spaces, no tabs

Yes, creating a project for this sounds like a good idea

**Comment by deinok** [comment]
@jedmao So what about creating a wiki or a file with this standards and conventions?

**Comment by ffes** [comment]
It would suggest a wiki page (or set of pages) for now. Not sure about the name though.

I think the links to the standard naming conventions mentioned here would be nice to include in the wiki as well.

Once that is done, we need to put it on the website as well. And maybe write a blog post about it.

Since every GitHub user can edit the wiki, this might result in vandalism or style wars. People can get fanatic about these kind of things.

**Comment by deinok** [comment]
@ffes I was thinking about a repository with this default file and a github pages that explain it, with links and references. What about this?

**Comment by deinok** [comment]
@jedmao Can you reopen the issue until a decision is made?

**Comment by ffes** [comment]
@deinok Re-opening is indeed a good idea

I have no strong opinion on repo vs wiki. 

"Everybody can edit the wiki" vs "handling pull request"
"Direct generated pages through github pages" vs "link to wiki page"

Both solutions have their advantages and disadvantages.

**Comment by deinok** [comment]
Well I vote "handling pull requests" and "Direct generated pages through github pages".

I think that allowing anybody to change the wiki is a bad idea. (War of code styles)

**Comment by xuhdev** [comment]
@jedmao @ffes @editorconfig/board-member How about we vote on this?

**Comment by jedmao** [comment]
I don't want a war, so I'm ok 👌 with the PR concept to keep subjective opinions from creeping in.

**Comment by sindresorhus** [comment]
PR

**Comment by xuhdev** [comment]
@sindresorhus @jedmao I've created a voting thread #299 . Please cast your formal votes there.

**Comment by ffes** [comment]
With the voting taking a direction, anyone a suggestion for a repo name?

And I just created a gist as a start that repo. All content based on this issue and personal experience.
https://gist.github.com/ffes/f950ced199ea6392589c4616c02c6be9

**Comment by deinok** [comment]
@ffes I also have a gist (https://gist.github.com/deinok/916e6c37c373f4af35d485b6f6f9d993)
Can you take everything you need from there. (Your gist is better than mine)

**Comment by deinok** [comment]
@ffes In java, the document that you link, it says:
`Four spaces should be used as the unit of indentation. The exact construction of the indentation (spaces vs. tabs) is unspecified. Tabs must be set exactly every 8 spaces (not 4).`

What i understand its that in some cases indentation is 8 spaces but tipicaly the common indentation is 4.
Ex:
if(something){
    doSomething();
}
someMethod(longExpression1, longExpression2, longExpression3, 
        longExpression4, longExpression5);

**Comment by xuhdev** [comment]
How about `editorconfig-default`?

**Comment by jedmao** [comment]
@xuhdev what if there are multiple defaults for the same language, based on multiple conventions?

**Comment by xuhdev** [comment]
@jedmao We just add comments on top of them. These files should not be used directly any way -- they are either for copy-and-paste purpose or used by some automation tools.

**Comment by jedmao** [comment]
I'm OK w/ `editorconfig-default`

**Comment by florianb** [comment]
I would really love to see machine-readable defaults. Multiple conventions could be defined in the subobjects as follows:

```yml
js:
  mozilla:
    # https://developer.mozilla.org/en-US/docs/Mozilla/Developer_guide/Coding_Style
    max_line_length: 80
    tab_width: 2
    indent_style: space
  jquery:
    # https://contribute.jquery.org/style-guide/js
    max_line_length: 100
    tab_width: 4
    indent_style: space
```

This example shows a yaml-file (since you're able to add comments).

I would expect that the website showing the defaults copy- and pastable is being generated from this file. Any tool should (if possible) directly rely on this file. This would ease the future maintenance of the defaults where PRs and Issues would document the why, while the standards-file  contains explicitly the what.

**Comment by xuhdev** [comment]
@florianb Good idea! I would even put the comments as an attribute.

**Comment by deinok** [comment]
@florianb @xuhdev Can editorconfig have this file format?

**Comment by florianb** [comment]
@xuhdev that's a good idea, too: having an attribute `reference` -- so website and tools could offer a direct link to the guidelines.. ❤️

**Comment by ffes** [comment]
I like the idea. The YAML syntax seems to fix this purpose nicely.

I am no Jekyll expert but after a quick peek at https://jekyllrb.com/docs/datafiles/ I **think** it should be possible to generate pages on the website with that.

Not sure though about the use of comment that way. I think special tags "_url" for the referenced url and "_pattern" for the regex file pattern would work better. If we go this path (and I think we should at least investigate it) we need to document the YAML properly so plugin maintainers know what to expect.

Having said that, I think most people want to look at the defaults website and just copy and paste it into their own files, so the final output (aka the website) is the very important.

Maybe the `editorconfig-default` repo should become a submodule of the website repo and just contain the YAML?

**Comment by florianb** [comment]
@deinok: i guess the file-format (in that case yaml) is pretty flexible. If we think about automating the generation of the site, i would even think about spitting out different file formats for the standard. F.e. the maintained default is `.yaml` and it is also provides as `exports/editorconfig-default.json` as well, to provide as most native input-file-formats as possible. 🤔

**Comment by deinok** [comment]
Ah, for the webpage. Seems Ok for me

**Comment by florianb** [comment]
@ffes 

> I am no Jekyll expert but after a quick peek at https://jekyllrb.com/docs/datafiles/ I think it should be possible to generate pages on the website with that.

You are right, i know it is easily possible.

> Not sure though about the use of comment that way. I think special tags "_url" for the referenced url and "_pattern" for the regex file pattern would work better. If we go this path (and I think we should at least investigate it) we need to document the YAML properly so plugin maintainers know what to expect.

Yaml also supports something like:

```yml
"*.js":
  mozilla:
    ... 
```

Introducing trailing underscores is possible, but i'd like to suggest avoiding them since you'd have to put it in quotes and the file would be read mainly by developers and machines anyways.

**Comment by ffes** [comment]
@florianb All valid points!

And do we choose YAML or JSON? They both seem to fit the purpose. I think we should choose one.

One disadvantage of JSON is that **officially** comments are not allowed, although I know most parsers don't mind. And since these data files need to maintained manually comments can be useful.

On the other hand, JSON is more commonly used.

In case we choose to submodule that repo: GitHub Pages is fine with that: https://help.github.com/articles/using-submodules-with-pages/

**Comment by florianb** [comment]
@ffes - the support of submodules is a great finding! ❤️ 

I personally would prefer not having any documentation in the default definition, even though i think it surely has advantages describing the "defaults" format in a documentation section.

I would leave the crosspilation-option open for future consideration because i could imagine developers would prefer being able to load the defaults file without adding additional dependencies like a Yaml library.

😄 of course, i would prefer being able to read the defaults in Js without the need to add an additional Yaml library.

**Comment by ffes** [comment]
@florianb So, I feel you have a preference for JSON :wink:

**Comment by johan** [comment]
@florianb's [example](https://github.com/editorconfig/editorconfig/issues/298#issuecomment-288640987) has useful data in thosee comments that I think could be useful (optional) reference links for each style for human-readable in-depths.

Even though `editorconfig` itself won't use those to configure your editor, I think there is value in encoding this data in a `.editorconfig` file, as humans occasionally read these too (say, `see_also`, `reference_url`, or something). And if there are or show up any cool editors that offer you syntactic rules help references online, we'd already have the data to feed that function for the relevant project.

Bonus: JSON would work, too, if we want to avoid yaml all together. (All sorts of atrocities are valid yaml, which not all naïve parsers grok.)

[I won't champion this proposal via formal process, but if someone likes it, feel free to run with it. :-]

**Comment by ffes** [comment]
Today I wanted to add a `.editorconfig` file to a small Python project, and I already missed the (yet to come) "Generate based on defaults" feature in my editor 😉

**Comment by ffes** [comment]
So what will be the preferred language to store those defaults? JSON or YAML?
I think JSON would be a better choice. It is better supported (often no need for extra libs) and more people are familiar with it.

And we need to work out a good schema. What will work for both situations, website generation and automatic processing by plugins.

**Comment by deinok** [comment]
+1 for JSON

**Comment by florianb** [comment]
👍 @ffes as you might already know i'd ❤️ using JSON for that.

I'd like to suggest that we define a minimium in the first step and then collect some experiences before evolving the definition.

💡 I put the following imagination to the discussion:

```json
{
"version":"0.1.0",
"matches": {
  "<<matcher>>": {
    "flavors": {
      "<<flavor>>": {
        "editorconfig": {
          "<<settings>>": "<<settings>>"
        },
        "description": "<<description>>",
        "reference": "<<reference>>"
      }
    }
  }
}
}
```
Where
- `matcher` stands for the default files-glob the section defines; the format should allow to use it to filter the defaults based on a list of files.
- `flavors` stands for the given flavor name; it should be used to let the user choose the preferred style; the using tool might omit showing it if there is only one flavor (f.e. `common`) for the section.
- `setting` are the editorconfig-settings for that flavor; they should be used to populate the editorconfig-file
- `description` contains the plain-text description of that flavor; it should be used as an introductory description of the flavor
- `reference` is the reference to the source of the flavor; it should lead to a authoritative description of the flavor

---
It might be a good idea to discuss the format in an extra thread.

**Comment by deinok** [comment]
Any update on this?

**Comment by xuhdev** [comment]
Someone has to work on it...

**Comment by xuhdev** [comment]
Another option is that we don't depend on jekyll: We can create a new repo that generates the HTML pages from the yaml files. The HTML page can be put under a different subdomain.

**Comment by ffes** [comment]
I think Jekyll is not the problem here. I just think your comment before that is more important. Somebody has to work on this.

Design a data structure to store everything, search for standards to describe, write the JSON (or YAML, although I think JSON is the one to choose), and once that is done we can see where and how we host it.

**Comment by deinok** [comment]
So, how is the state of this?
The initial idea is to have a default editorconfig file to make all new projects follow the same conventions.

**Comment by mcandre** [comment]
A sane default .editorconfig, overridden by local settings would be really helpful, and reduce the amount of manual configuration required!

**charset=utf-8** should be the default for the vast majority of text files.

**insert_final_newline = true** is generally recommended, and helps `cat` to look right.

**trim_trailing_whitespace = true** is common sense for most source code, though some ASCII .txt files may prefer to maintain trailing space for artistic purposes.

**Make** and **Go** require hard tabs at a lexer level, so soft tabs shouldn't even be allowed for these files. Likewise, F# requires soft tabs, so EditorConfig should ideally reject a hard tab configuration being applied to F# files.

**PEP8** style should be the default for Python code, it's actually a language-wide standard.

**bash8 AKA bashate** applies a PEP8-like style for shell scripts. OpenStack is recommending these settings for hardcore, robust shell scripts. Would be helpful to accordingly apply bashate settings as the default for shell scripts, unless otherwise overridden.

**`*.min.*`** (minified files) should not be stylized at all, as these are machine-generated.

**LF** line endings should be default for most source code, with exceptions being .NET/.bat/.cmd and other Windows-centric files. Windows users may want .txt files to remain CRLF, for Notepad compatibility. But LF is generally going to lead to a better experience in modern development setups.

JavaScript code often follows the Google style, or else the Mozilla style. Google style may be a better choice of default, as it makes the smart move of preferring single quotes in JS strings, in order to nest more easily within HTML code.

Regardless of the particular style choices, adopting SOME kind of default on a per-language basis would substantially relieve the burden of manual configuration, and help steer users towards reasonably good looking source code.

**Comment by ffes** [comment]
Last night I have been putting some time in this and I think I have a decent working solution.

I  [attached a zip](https://github.com/editorconfig/editorconfig/files/1460594/defaults.zip) with these 3 files:

- defaults.json
- index.html
- README.md

The layout of `defaults.json` is based on the suggestion done by @florianb above. The content is largely based on my own gist mentioned here as well. This file should go in a separate repo named `editorconfig-default` as discussed in this issue. I also copied @florianb text to describe the format and put it in `README.md` that will the start for the one to go into that repo.

To generate the page in the website, for now I put `defaults.json` in the `_data` directory. This way Jekyll can read the data and and generate the webpage from it.

This screenshot gives an impression what the generated page will look like.

![screenshot](https://user-images.githubusercontent.com/3225656/32649316-03702342-c5fa-11e7-9bd9-3efac861bb5a.png)

So I think the next step should be create that new repo named `editorconfig-default`. I don't have permissions to do that in our org. But I can create one in my own account and we can transfer that later. Once that repo is created, I can add the page to the website repo, using a `git submodule`.

After that we can create a blogpost about this feature describing it.

Hopefully plugin developers will start using the `defaults.json` that is in that repo (ideally open the raw file from github) to add a feature to their plugin to add snippets to an `.editorconfig` file. That would be a major win!

First I want to get the infrastructure right before we add extra languages and flavors to the `defaults.json` file.

Did I overlook something? Anybody any other suggestion?

**Comment by deinok** [comment]
Hey, i like the idea this defaults.json seems quite useful to change quickly of one standard to another.
But my suggestion is that we should give a priority to `flavors`.

Example JavaScript: Mozilla have more 'priority' over jQuery, then mozilla will be used by default and only change if i override .editorconfig to jquery or use myself [*.js] config

Some langs only have one convention like Python, C#, etc...

**Comment by ffes** [comment]
@deinok I think your suggestion to add a `priority` is a good one, but we need to be careful what we give a high priority. We don't want to start a religious war on that.

For languages like Python or C#, we can safely choose a high priority, but JavaScript is bit more difficult. I just looked for a Node.js coding style and found [Felix Node.js Style Guide](https://github.com/felixge/node-style-guide). That *seems* to be what most node devs use (I'm not a node dev) used. Although it is very similar to the Mozilla, which one do you choose as highest prio? Both?

**Comment by deinok** [comment]
@ffes I'm agree with don't start a war :(
The thing i say about JavaScript was only an example of two code conventions. 
I think that we should use by default the convention of the language over organizations conventions.

Example: Use EcmaScript code convention (if it exist) over Mozilla, JQuery and Node
(I'm not a web developer so i can be missign something)

**Comment by ffes** [comment]
One thing I noticed that I put `true` and `false` in the json in the zip as strings, but they should have been real boolean values obviously.

**Comment by deinok** [comment]
Yeah, we could start working in a JSON-Schema in order to define all this things and the specification of this.

**Comment by ffes** [comment]
Validation is always a good thing. But I leave that up to others for now. I have never done that for JSON before.

**Comment by deinok** [comment]
I can work on it this weekend if you want.

**Comment by ffes** [comment]
@deinok That would be great!

**Comment by deinok** [comment]
@ffes Okey, it will be ready for the end of the weekend ;)

**Comment by deinok** [comment]
@ffes As we talked, I have done the JSON-Schema following the `draft-06`.
The schema can be improved in several ways. And also I have detected some mistakes with the format of the JSON.

Do you thing we should move this to his own repo? Something like https://github.com/editorconfig/editorconfig-defaults

[editorconfig-defaults.zip](https://github.com/editorconfig/editorconfig/files/1462203/editorconfig-defaults.zip)

**Comment by mcandre** [comment]
Final EOL should be enabled, for `cat` support and to ensure that C headers are properly handled. Technically, a final end of line character is also required in order to for a text file to be POSIX compliant.

I'd recommend that all text files default to LF, except specifically Microsoft Windows development files (bat, com, vbs, cs, fs, fsi, and so on).

**Comment by ffes** [comment]
@deinok Thanks! I will look at your files ASAP. I thought it was going to be an easy going weekend, but turns out it is not (for good reasons) \o/

@mcandre In general you have valid points, although some people might say those are your personal preferences. So we have to be careful not start a large discussion (or even religious war) on which settings are better. See for instance the discussion in #226, but we have had that at other times as well. With the flavors we can deal with that reasonably well, I think.

As I said before. I first want to get the infrastructure right. That will be my main focus at the moment. After that we can add extra defaults in the dedicated repo.

**Comment by deinok** [comment]
@mcandre im totally agree with @ffes.
It have sense to default EOF, and for me LF its problematic.
For this step, its more important to define the JSON structure than the default values itself.

**Comment by florianb** [comment]
@deinok i'd like to avoid introducing priorities of any kind - in my opinion those would add another dimension to the already existing prioritization of editorconfig-settings. I could however imagine to default to the most common and/or idiomatic style (by usage-counts).

I'd prefer extending the evaluation chain of the editorconfig-files. Currently ecfg-files are evaluated up to the root, it would probably be a good idea to add user-specific and a system-wide ecfg-location to recognize (i know this has been controversially discussed but as complexity gains we should discuss it again).

In addition to that i think it is necessary to introduce the possibility to choose a flavor explicitly by having a new setting like `default = js.airbnb`.

I'm looking forward contributing to the dedicated repository. ❤️

**Comment by ffes** [comment]
@deinok I took a quick look. How did you use that validator? I assume you used node to validate, but there are so many json validation packages. Could you upload your package.json here are well?

**Comment by deinok** [comment]
@ffes JSON-Schema is not language dependent. You can use this webpage for quick validation https://www.jsonschemavalidator.net/ (Put the editorconfig-defaults.schema.json on the left and any editorconfig-defaults.json on the right)

**Comment by deinok** [comment]
@florianb well, priority it could be not a good example,  in the flavours of a language it should have some kind of `default:true` to indicate that in case that it is not explicit indicated that flavour is going to be used.
And yes, in my opinion the .editorconfig file should have a way to indicate that "prefered flavour".

Another question is:
This editorconfig-defaults.json should be touched by the users or should be integrated with IDEs, or the two things.

**Comment by ffes** [comment]
Giving it a second thought, you are right that `priority` is not the best way to identify the official standards like with YAML and Python's PEP8.

So for now I want to leave that problem/subject AS-IS. I will add an issue to the new repo when it is there. I will stick with @florianb suggested file format for now.

**Comment by deinok** [comment]
Im agree all this concepts and discusions can be handled in separated issues in the future repo.

The important thing is that seems like we are going to have some kind of default editorconfig. :)

**Comment by ffes** [comment]
> This editorconfig-defaults.json should be touched by the users or should be integrated with IDEs, or the two things.

Not sure what you mean by `touched by the users`, but I think that in general the file is meant to be used for machine reading. Two thing I can see it being used for:

- From that we can generate a dedicated page on the main EditorConfig website.
- Plugin-devs can use this file to improve their plugin to allow snippets be inserted in a new or existing `.editorconfig` file. For instance the vscode plugin has an option to create a new `.editorconfig` file. That feature could use this file.

I'm sure others will come up with even better use of this file, but those two would already be great to have.

**Comment by deinok** [comment]
@ffes I mean, if a normal user will be able to modify the defaults file. 
Needs discussion

**Comment by ffes** [comment]
See https://github.com/ffes/editorconfig-defaults

This new repo should be transferred to the editorconfig organization. I will open a separate issue for that unless @treyhunner or @xuhdev have transferred it already.

I will close this issue, so we can continue the discussions there.

I will open a couple of issues in the new repo, with various points that still need to be addressed.

**Comment by deinok** [comment]
@ffes Okey, wait until the repo is transfered to close this issue.
Also if you could give me write rights to that repo i will apreciate it.

**Comment by xuhdev** [comment]
@ffes Thanks, can you initiate the transfer?

**Comment by ffes** [comment]
@xuhdev No, I can't. I don't have the permission:

>You don’t have the permission to create repositories on editorconfig

**Comment by xuhdev** [comment]
@ffes Now I remembered how this should proceed. You need to give me the admin permission of the repo and I can transfer the ownership.

**Comment by ffes** [comment]
@xuhdev I don't see how I can give you admin persmissions to my repo, but I added you as collaborator. You should have seen the invite.

**Comment by ffes** [comment]
@deinok 

> Also if you could give me write rights to that repo i will apreciate it.

Since the repo is going to be transferred, the permissions will be enforced through the organization. And I don't control those.

**Comment by xuhdev** [comment]
@ffes It looks like you need to transfer the repo to me first... GitHub seems having made changes (repos under users cannot be co-owned).

**Comment by xuhdev** [comment]
Transfer done: https://github.com/editorconfig/editorconfig-defaults

**Comment by ffes** [comment]
Thanks a lot! That closes this issue.

**Comment by mcandre** [comment]
Just to add fuel to the fire, here is the configuration I am using for most of my projects:

https://github.com/mcandre/dotfiles/blob/master/.editorconfig

I elected to start using four spaces of indentation for 99% of my code, since shell and python were already standardizing on that style. Note the unfortunate use of `unset` rather than `none`, as many CLI tools fail to match the EditorConfig spec!

**Comment by RehanSaeed** [comment]
![logo](https://user-images.githubusercontent.com/2270364/62460937-4ea7f000-b77b-11e9-89b8-0abca26bfd4f.png)
A very generic .editorconfig file supporting .NET, C#, VB and web technologies.
https://github.com/RehanSaeed/EditorConfig

Currently has 151 stars, so used by a fair number of people in the .NET community.
